### ISSUE TITLE
[TF 2.0 API Docs] `tf.keras.callbacks` (sub)classes

### ISSUE THREAD

**Issue by 8bitmp3** [tensorflow/tensorflow#29958]
TensorFlow version: 2.0 (beta1)

#### (Sub)classes of `tf.keras.callbacks`:
`BaseLogger`, `History`, `Callback`, `CSVLogger`, `ModelCheckpoint`, `ProgbarLogger`, `RemoteMonitor`, `TensorBoard` and `TerminateOnNaN`:

#### Summary:
Since `tf.keras.callbacks` are important for monitoring models during training, the API docs require extra attention imo.

Examples - to add - see below
Descriptions - to be defined better - see below (especially the `Callback` custom class)
Returns/raises - to add for better UX when needed

#### Suggested improvements: 

- _Missing examples_: one example per (sub)class would be enough for good UX and a link to a tutorial presents an extra step for a user. E.g. a short example inside the docs similar to the ones in `EarlyStopping` ([link](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/EarlyStopping)), `ReduceLROnPlateau` ([link](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/ReduceLROnPlateau)) or `LambdaCallback` ([link](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/LambdaCallback))

Note: `ModelCheckpoint` and `TensorBoard` have links to tutorials which have examples. `BaseLogger` and `History` are applied by default but that may not help understand them better.

Example of a `Custom` callback from @fchollet Deep Learning with Python book:
```python
class ActivationLogger(keras.callbacks.Callback):
    def set_model(self, model):
        self.model = model # Called by the parent model before training, to inform the callback of what model will be calling it
        layer_outputs = [layer.output for layer in model.layers]
        self.activations_model = keras.models.Model(model.input, layer_outputs) # Model instance that returns the activations of every layer

    def on_epoch_end(self, epoch, logs=None):
        if self.validation_data is None:
                raise RuntimeError('Requires validation_data.')
        validation_sample = self.validation_data[0][0:1] # Obtains the first input sample of the validation data
        activations = self.activations_model.predict(validation_sample)
        f = open('activations_at_epoch_' + str(epoch) + '.npz', 'w') # Saves arrays to disk
        np.savez(f, activations)
        f.close()
```

Example of a `TensorBoard` callback from @fchollet Deep Learning with Python book:
```python
callbacks = [
    keras.callbacks.TensorBoard(
        log_dir='my_log_dir', # Location of log files
        histogram_freq=1, # Records activation histogram every 1 epoch
        embeddings_freq=1, # Records embedding data every 1 epoch
    )
]

history = model.fit(x_train, y_train, 
                epochs=20, 
                batch_size=128, 
                validation_split=0.2, 
                callbacks=callbacks)

# Browse to http://localhost:6006 and look at your model training
...
```

- _Descriptions_: to be defined better if needed. Recommend to use the following Medium post  which has quite decent descriptions of each `callback`: https://medium.com/singlestone/keras-callbacks-monitor-and-improve-your-deep-learning-205a8a27e91c

Note: Mention in the `EarlyStopping` and `ModelCheckpoint` descriptions that they are/should be both typically used together (see `callbacks_list` example from @fchollet with 2 callbacks passed into `model.fit` below) to stop training when improvement stops and save the current best model during training (`save_best_only=True`):

```python
# A list of 2 or more callbacks that can be passed into `model.fit`
callbacks_list = [
        keras.callbacks.EarlyStopping(
                monitor='acc',
                patience=1,
        ),
        
        keras.callbacks.ModelCheckpoint( # Saves the current weights after every epoch
                filepath='my_model.h5',
                monitor='val_loss',
                save_best_only=True, # These two arguments mean you won’t overwrite the model file unless val_loss has improved
        )
]

model.compile(optimizer='rmsprop',
                loss='binary_crossentropy',
                metrics=['acc'])

model.fit(x, y,
                epochs=10,
                batch_size=32,
                callbacks=callbacks_list,
                validation_data=(x_val, y_val)
                )
```

- _Returns/Raises_: to be defined/defined better if needed

#### Doc links:
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/BaseLogger
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/History
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/Callback
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/CSVLogger
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/ModelCheckpoint
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/ProgbarLogger
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/RemoteMonitor
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/TensorBoard
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/TerminateOnNaN

**Comment by PigApple** [comment]
@8bitmp3, is there any possible way to dump model gradients and intermediate layer output through callback?

**Comment by hassanharoon321** [comment]
how to do classificaiton using Fashion MNIST, a data set containing items of clothing. There's another, similar dataset called MNIST which has items of handwriting -- the digits 0 through 9.

Write an MNIST classifier that trains to 99% accuracy or above, and does it without a fixed number of epochs -- i.e. you should stop training once you reach that level of accuracy.

Some notes:

It should succeed in less than 10 epochs, so it is okay to change epochs= to 10, but nothing larger
When it reaches 99% or greater it should print out the string "Reached 99% accuracy so cancelling training!"
If you add any additional variables, make sure you use the same names as the ones used in the class


can you solve this

**Comment by hassanharoon321** [comment]
import tensorflow as tf
from os import path, getcwd, chdir

# DO NOT CHANGE THE LINE BELOW. If you are developing in a local
# environment, then grab mnist.npz from the Coursera Jupyter Notebook
# and place it inside a local folder and edit the path to that location
path = f"{getcwd()}/../tmp2/mnist.npz"


# GRADED FUNCTION: train_mnist
class myCallback(tf.keras.callbacks.Callback):
    def train_mnist(self, epoch, logs={}):
        # Please write your code only where you are indicated.
        # please do not remove # model fitting inline comments.
        if(logs.get('acc')>0.99):
              print("\nReached 99% accuracy so cancelling training!")
        self.model.stop_training = True
    mnist = tf.keras.datasets.mnist
    (x_train, y_train),(x_test, y_test) = mnist.load_data(path=path)
    # YOUR CODE SHOULD START HERE
    x_train, x_test = x_train / 255.0, x_test / 255.0
    # YOUR CODE SHOULD END HERE
    model = tf.keras.models.Sequential([
        # YOUR CODE SHOULD START HERE
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(512, activation=tf.nn.relu),
    tf.keras.layers.Dense(10, activation=tf.nn.softmax)
        # YOUR CODE SHOULD END HERE
    ])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
    
    # model fitting
history = model.fit(x_train, y_train, epochs=10, callbacks=[callbacks])
    # model fitting
return history.epoch, history.history['acc'][-1]

**Comment by Krishnarohith10** [comment]
I am using tensorflow==2.1 and when I ran this algorithm I got error saying this:
![1](https://user-images.githubusercontent.com/44919399/73130987-fca89e80-4028-11ea-85d7-b7e25c34e621.jpg)
Briefly it says that logs.get('acc') is Nonetype and 0.99 is float so, '>' this cannot compare.
Can someone help me?

**Comment by mithuntp-online** [comment]
> I am using tensorflow==2.1 and when I ran this algorithm I got error saying this:
> ![1](https://user-images.githubusercontent.com/44919399/73130987-fca89e80-4028-11ea-85d7-b7e25c34e621.jpg)
> Briefly it says that logs.get('acc') is Nonetype and 0.99 is float so, '>' this cannot compare.
> Can someone help me?

Same Issue here, any solution found?

**Comment by Krishnarohith10** [comment]
For tensorflow==2.1 it is not log.get('acc') instead try log.get('accuracy'). It should work.

**Comment by mithuntp-online** [comment]
> I am using tensorflow==2.1 and when I ran this algorithm I got error saying this:
> ![1](https://user-images.githubusercontent.com/44919399/73130987-fca89e80-4028-11ea-85d7-b7e25c34e621.jpg)
> Briefly it says that logs.get('acc') is Nonetype and 0.99 is float so, '>' this cannot compare.
> Can someone help me?

Hi..
Use if(logs.get('acc')>0.90):

This worked for me

**Comment by Krishnarohith10** [comment]
Hello..
For me it didn't, then I used 'accuracy' instead of 'acc', this worked for me.

**Comment by joyanta180199** [comment]
For tensorflow version 1.14.0, try log.get('acc')
And as per the comments, if tf version is more than 2, then try log.get('accuracy')
